### PR TITLE
Chapter 11: Account activation

### PR DESCRIPTION
**Pull request by rashidfirhat98**

### CODE REVIEW

**Comment by rashidfirhat98** [comment]
ready

File: app/models/user.rb [hunk 1]
```diff
@@ -46,9 +47,28 @@ def authenticated? remember_token
     BCrypt::Password.new remember_digest.is_password? remember_token
   end
 
+  def authenticated? attribute, token
+    digest = send "#{attribute}_digest"
+    return false if digest.nil?
+    BCrypt::Password.new(digest).is_password? token
```
Comment (saobv-2223) [inline app/models/user.rb]:
```suggestion
    return false unless digest
    
    BCrypt::Password.new(digest).is_password? token
```

File: app/models/user.rb [hunk 2]
```diff
@@ -46,9 +47,28 @@ def authenticated? remember_token
...
   private
 
   def downcase_email
-    email.downcase!
+    self.email.downcase!
+  end
+
+  def create_activation_digest
+    self.activation_token = User.new_token
+    self.activation_digest = User.digest(activation_token)
```
Comment (saobv-2223) [inline app/models/user.rb]: remove ()

File: config/environments/development.rb [hunk 3]
```diff
@@ -31,9 +31,20 @@
...
-  config.action_mailer.raise_delivery_errors = false
-
+  # Mailer 
+  config.action_mailer.raise_delivery_errors = true
+  config.action_mailer.delivery_method = :test
+  host = "localhost"
+  config.action_mailer.default_url_options = { host: host, 
+    port:3000, protocol: 'http' }
+  ActionMailer::Base.smtp_settings = {
+    address: 'smtp.gmail.com',
```
Comment (saobv-2223) [inline config/environments/development.rb]: using "

Comment (rashidfirhat98) [reply config/environments/development.rb]: done

**Comment by rashidfirhat98** [comment]
ready

File: app/mailers/user_mailer.rb [hunk 4]
```diff
@@ -0,0 +1,23 @@
+class UserMailer < ApplicationMailer
+
+  # Subject can be set in your I18n file at config/locales/en.yml
+  # with the following lookup:
+  #
+  #   en.user_mailer.account_activation.subject
+  #
```
Comment (saobv-2223) [inline app/mailers/user_mailer.rb]: remove all the comment lines

File: app/models/user.rb [hunk 5]
```diff
@@ -46,9 +47,29 @@ def authenticated? remember_token
     BCrypt::Password.new remember_digest.is_password? remember_token
   end
 
+  def authenticated? attribute, token
+    digest = send "#{attribute}_digest"
+    return false if digest.nil?
```
Comment (saobv-2223) [inline app/models/user.rb]:
```suggestion
    return false unless digest
   
```

Comment (rashidfirhat98) [reply app/models/user.rb]: done

**Comment by rashidfirhat98** [comment]
ready

**Comment by ngocvt-0484** [comment]
merged

**Pull request merged**